File: sdk/packages/python/iii/src/iii/state.py
"""State management types and interface for the III SDK.

Mirrors the Node SDK ``iii-sdk/state`` subpath export.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from enum import Enum
from typing import Any, Generic, TypeVar

from iii_helpers.stream import UpdateOp
from pydantic import BaseModel

TData = TypeVar("TData")


class StateGetInput(BaseModel):
    """Input for retrieving a state value."""

    scope: str
    key: str


class StateSetInput(BaseModel):
    """Input for setting a state value."""

    scope: str
    key: str
    value: Any


class StateDeleteInput(BaseModel):
    """Input for deleting a state value."""

    scope: str
    key: str


class StateListInput(BaseModel):
    """Input for listing all values in a state scope."""

    scope: str


class StateUpdateInput(BaseModel):
    """Input for atomically updating a state value."""

    scope: str
    key: str
    ops: list[UpdateOp]


class StateSetResult(BaseModel, Generic[TData]):
    """Result of a state set operation."""

    old_value: TData | None = None
    new_value: TData


class StateUpdateResult(BaseModel, Generic[TData]):
    """Result of a state update operation."""

    old_value: TData | None = None
    new_value: TData


class StateDeleteResult(BaseModel):
    """Result of a state delete operation."""

    old_value: Any | None = None


class StateEventType(str, Enum):
    """Types of state change events."""

    CREATED = "state:created"
    UPDATED = "state:updated"
    DELETED = "state:deleted"


class StateEventData(BaseModel, Generic[TData]):
    """Payload for state change events."""

    type: str = "state"
    event_type: StateEventType
    scope: str
    key: str
    old_value: TData | None = None
    new_value: TData | None = None


class IState(ABC, Generic[TData]):
    """Abstract interface for state management operations."""

    @abstractmethod
    async def get(self, input: StateGetInput) -> TData | None:
        """Retrieve a value by scope and key."""
        ...

    @abstractmethod
    async def set(self, input: StateSetInput) -> StateSetResult[TData] | None:
        """Set (create or overwrite) a state value."""
        ...

    @abstractmethod
    async def delete(self, input: StateDeleteInput) -> StateDeleteResult:
        """Delete a state value."""
        ...

    @abstractmethod
    async def list(self, input: StateListInput) -> list[TData]:
        """List all values in a scope."""
        ...

    @abstractmethod
    async def update(self, input: StateUpdateInput) -> StateUpdateResult[TData] | None:
        """Apply atomic update operations to a state value."""
        ...

File: sdk/packages/node/iii/src/state.ts
import type { UpdateOp, UpdateOpError } from '@iii-dev/helpers/stream'

/** Input for retrieving a state value. */
export type StateGetInput = {
  /** State scope (namespace). */
  scope: string
  /** Key within the scope. */
  key: string
}

/** Input for setting a state value. */
export type StateSetInput = {
  /** State scope (namespace). */
  scope: string
  /** Key within the scope. */
  key: string
  /** Value to store. */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  value: any
}

/** Input for deleting a state value. */
export type StateDeleteInput = {
  /** State scope (namespace). */
  scope: string
  /** Key within the scope. */
  key: string
}

/** Result of a state delete operation. */
export type StateDeleteResult = {
  /** Previous value (if it existed). */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  old_value?: any
}

/** Input for listing all values in a state scope. */
export type StateListInput = {
  /** State scope (namespace). */
  scope: string
}

/** Result of a state set operation. */
export type StateSetResult<TData> = {
  /** Previous value (if it existed). */
  old_value?: TData
  /** New value that was stored. */
  new_value: TData
}

/** Result of a state update operation. */
export type StateUpdateResult<TData> = {
  /** Previous value (if it existed). */
  old_value?: TData
  /** New value after the update. */
  new_value: TData
  /**
   * Per-op errors. Currently emitted only by the `merge` op when input
   * violates the validation bounds. See {@link UpdateOpError} and the
   * `UpdateMerge` JSDoc in `./stream` for the error codes. Field is
   * omitted from the JSON wire when empty.
   */
  errors?: UpdateOpError[]
}

/** Input for atomically updating a state value. */
export type StateUpdateInput = {
  /** State scope (namespace). */
  scope: string
  /** Key within the scope. */
  key: string
  /** Ordered list of update operations to apply atomically. */
  ops: UpdateOp[]
}

/** Types of state change events. */
export enum StateEventType {
  Created = 'state:created',
  Updated = 'state:updated',
  Deleted = 'state:deleted',
}

/** Payload for state change events. */
// biome-ignore lint/suspicious/noExplicitAny: any is fine here
export interface StateEventData<TData = any> {
  type: 'state'
  /** Type of state change. */
  event_type: StateEventType
  /** State scope (namespace). */
  scope: string
  /** Key within the scope. */
  key: string
  /** Previous value (for update/delete events). */
  old_value?: TData
  /** New value (for create/update events). */
  new_value?: TData
}

/**
 * Interface for state management operations. Available via the `iii-sdk/state`
 * subpath export.
 */
export interface IState {
  /** Retrieve a value by scope and key. */
  get<TData>(input: StateGetInput): Promise<TData | null>
  /** Set (create or overwrite) a state value. */
  set<TData>(input: StateSetInput): Promise<StateSetResult<TData> | null>
  /** Delete a state value. */
  delete(input: StateDeleteInput): Promise<StateDeleteResult>
  /** List all values in a scope. */
  list<TData>(input: StateListInput): Promise<TData[]>
  /** Apply atomic update operations to a state value. */
  update<TData>(input: StateUpdateInput): Promise<StateUpdateResult<TData> | null>
}
